### PR TITLE
opensc-notify: register a desktop menu entry

### LINKED ISSUE #1186: Running without $DISPLAY shows error from notification code
### Problem Description

Running `pkcs11-tool` on machine without `$DISPLAY`/without X11 server running throws the following bogus errors, while attempting to show the smartcard notifications:
```
Error connecting: Cannot autolaunch D-Bus without X11 $DISPLAY
```
### Proposed Resolution
On Linux, the notification code should not attempt to show notification if there is no `$DISLPAY` environment variable set.

### Steps to reproduce
Connect remotely to machine with smart card and without display and list the PKCS#11 objects (without [PCSC access control](https://access.redhat.com/blogs/766093/posts/1976313)):
```
./pkcs11-tool -O
```
Unfortunately it is not possible to reproduce the issue by unsetting `$DISPLAY`.

### PR DESCRIPTION
**Pull request by frankmorgner**
- fixes showing notifications in gnome-shell via gio2
- removes gdbus interface for notifications
- fixes #1186

##### Checklist

- [x] Tested with the following card: sc-hsm
	- [x] tested PKCS#11

### CODE REVIEW

**Comment by frankmorgner** [comment]
@Jakuje could you check whether this fixes your problem?

**Comment by Jakuje** [comment]
No. With this patch, it fails the show the notification altogether with the following error even if I have a display:
```
(process:18872): GLib-GIO-CRITICAL **: g_application_send_notification: assertion 'G_IS_APPLICATION (application)' failed
```
But I was testing just from the local build without installation (though the current master work with this setup properly).

**Comment by frankmorgner** [comment]
I forgot to commit a change that was present locally. Sorry, for the inconvenience.

**Comment by Jakuje** [comment]
This does not throw error messages, but does not show the notification either in GNOME as it was showing before. Does it require to install the desktop entry somewhere?

**Comment by frankmorgner** [comment]
Yes, gnome shell requires you to put it into one of `$XDG_DATA_DIRS`'s `applications` directories (e.g. putting `org.opensc-project.opensc-notify.desktop` into `~/.local/share/applications/` will do)

File: src/tools/Makefile.am [hunk 1]
```diff
@@ -178,5 +179,11 @@ gids_tool_SOURCES += versioninfo-tools.rc
 opensc_notify_SOURCES += versioninfo-opensc-notify.rc
 endif
 
+applicationsdir = $(datadir)/applications
+applications_DATA = org.opensc-project.opensc-notify.desktop
+
+org.opensc-project.opensc-notify.desktop: org.opensc-project.opensc-notify.desktop
```
Comment (Jakuje) [inline src/tools/Makefile.am]: This should say (adding `.in` to the end):
```
org.opensc-project.opensc-notify.desktop: org.opensc-project.opensc-notify.desktop.in
```

**Comment by Jakuje** [comment]
For some reason, I don't have `~/.local/share` in `$XDG_DATA_DIRS`. But even after putting it in into `/usr/share/applications/` and after running `desktop-file-validate` with this desktop file I do not see any notification nor any error. Hard to say what could go wrong.

Looking into the configure log, it looks like everything is defined as expected. Not sure what else I could check or if the debug log would be useful.

**Comment by frankmorgner** [comment]
Strange... have you tried a different desktop environment? AFAIK, gnome shell is the only one that requires you to have a desktop entry.

**Comment by Jakuje** [comment]
I don't have any other DE installed and set up to try at this moment.

I tried to intercept the notification using DBus, but I did not see them either according to the following post but nothing happens (for example using `pkcs11-tool`):

https://askubuntu.com/questions/105566/is-there-a-way-to-view-notification-history/352125#352125

On the other hand, I can see the same failures
for other tools using `libopensc`, for example if I try  `tests/pintest`, which is not improvement either (before applying this patch, it was showing the notification properly even from here).
```
(process:20090): GLib-GIO-CRITICAL **: g_application_send_notification: assertion 'G_IS_APPLICATION (application)' failed
```
There should be some check, that the notify is initialized, before trying to show notification, should not it?

I even tried to run the example notification program using GIO2 from archlinux, but still no notification:

https://wiki.archlinux.org/index.php/Desktop_notifications

Though I am able to get the notifications using `libnofity` example program and I am getting the notifications from other programs on my desktop. I am probably out of ideas where to search or what to try so at least these ideas so far.

File: src/tools/Makefile.am (hunk 1 above)
Comment (frankmorgner) [reply src/tools/Makefile.am]: done, thanks

**Comment by frankmorgner** [comment]
I had the same problem for quite some time, that's why I've implemented the `gdbus` workaround. But as explained above, I solved the problem with a desktop file for gnome-shell. I hope that the default install location I've used will work on other distributions, too...

I've added the check on whether the notifications are initialized.

**Comment by frankmorgner** [comment]
I've noticed that I've got the following in my shell startup configuration:
```
export XDG_DATA_DIRS=$XDG_DATA_DIRS:$HOME/.local/share
```
Setting this may let xdg find your desktop file.

**Pull request merged**